File: LLD/Design-Patterns/Iterator/code/iterator.js
/**
 * Iterator Pattern — a Playlist exposes a PlaylistIterator instead of its internal array.
 * Run: node iterator.js
 */

class Playlist {
  constructor() {
    this.songs = [];
  }

  addSong(song) {
    this.songs.push(song);
  }

  createIterator() {
    return new PlaylistIterator(this.songs);
  }
}

class PlaylistIterator {
  constructor(songs) {
    this.songs = songs;
    this.position = 0;
  }

  hasNext() {
    return this.position < this.songs.length;
  }

  next() {
    const song = this.songs[this.position];
    this.position += 1;
    return song;
  }
}

const playlist = new Playlist();
playlist.addSong("Clair de Lune");
playlist.addSong("Gymnopedie No.1");
playlist.addSong("River Flows in You");

const iterator = playlist.createIterator();

console.log("Playing playlist:");
while (iterator.hasNext()) {
  console.log(`Now playing: ${iterator.next()}`);
}

module.exports = { Playlist, PlaylistIterator };

File: v1/LLD/Design-Patterns/Iterator/code/Iterator.cpp
// Iterator Pattern — a Playlist exposes a PlaylistIterator instead of its internal vector.
// Compile: g++ -std=c++14 Iterator.cpp -o iterator
// Run:     ./iterator

#include <iostream>
#include <string>
#include <vector>

class PlaylistIterator {
public:
    explicit PlaylistIterator(const std::vector<std::string>& songs) : songs_(songs), position_(0) {}

    bool hasNext() const {
        return position_ < songs_.size();
    }

    std::string next() {
        std::string song = songs_[position_];
        position_++;
        return song;
    }

private:
    const std::vector<std::string>& songs_;
    size_t position_;
};

class Playlist {
public:
    void addSong(const std::string& song) {
        songs_.push_back(song);
    }

    PlaylistIterator createIterator() const {
        return PlaylistIterator(songs_);
    }

private:
    std::vector<std::string> songs_;
};

int main() {
    Playlist playlist;
    playlist.addSong("Clair de Lune");
    playlist.addSong("Gymnopedie No.1");
    playlist.addSong("River Flows in You");

    PlaylistIterator iterator = playlist.createIterator();

    std::cout << "Playing playlist:" << std::endl;
    while (iterator.hasNext()) {
        std::cout << "Now playing: " << iterator.next() << std::endl;
    }

    return 0;
}
